cpp/src/codegen/arrow_compute/ext/array_item_index.h

```diff
@@ -31,6 +31,7 @@ struct ArrayItemIndex {
 struct ArrayItemIndexS {
   uint16_t id = 0;
   uint16_t array_id = 0;
+  uint64_t prefix = 0;
```

Comment:
this will increase the memory usage by 3x, can we dynamically projection the prefix?